### PR TITLE
[#59] Implement ActionSelector

### LINKED ISSUE #59: Action Selector
Fill missing implementation of the ActionSelector class provided [here](https://github.com/cyfronet-fid/recommender-system/blob/a787b6f505f6b4cfe34a43cbbaf4774226536846/recommender/engine/rl_agent/action_selector.py).

### PR DESCRIPTION
**Pull request by wujuu**
Closes #59

### CODE REVIEW

File: recommender/engine/rl_agent/action_selector.py [hunk 1]
```diff
@@ -18,24 +50,64 @@ def __call__(self, weights: torch.Tensor, user: User, search_data: SearchData) -
             weights: Weights returned by an actor model, tensor of shape [K, SE], where:
                 - K is the number of services required for the recommendation
                 - SE is a service content tensor embedding dim
-            user: User for whom the recommendation is generated. User's accessed_services is used for narrowing the itemspace down.
```
Comment (JanKapala) [inline recommender/engine/rl_agent/action_selector.py:21]: To obtain all ordered services, both user.accessed_services and user actions (of the same user) should be used (there can be iinformation about ordering that is not avalilable in the user.accessed_services - that comes from DB dump). Appropriate function will be implemented in [Services History Generator](https://github.com/cyfronet-fid/recommender-system/blob/e8ade68838aa3d5a11e7606acddab26539e50af5/recommender/services/services_history_generator.py) according to this issue #83

File: recommender/engine/rl_agent/action_selector.py [hunk 2]
```diff
@@ -18,24 +50,64 @@ def __call__(self, weights: torch.Tensor, user: User, search_data: SearchData) -
...
 
-        # TODO: implement
-        pass
+        all_services = Service.objects.order_by("id")
+        self.I = len(all_services)
+        self.index_id_map = pd.DataFrame(
+            sorted(all_services.distinct("id")), columns=["id"]
+        )
+
+        service_one_hot_tensors = torch.Tensor([s.tensor for s in all_services])
```
Comment (JanKapala) [inline recommender/engine/rl_agent/action_selector.py:108]: Maybe it would be more clear to use torch.stack method.

**Review by JanKapala — CHANGES_REQUESTED** [review]
Nice work. I marked few important details.

File: recommender/engine/rl_agent/action_selector.py (hunk 2 above)
Comment (wujuu) [reply recommender/engine/rl_agent/action_selector.py:108]: this method will be purged anyway during refactoring

File: recommender/engine/rl_agent/action_selector.py (hunk 1 above)
Comment (wujuu) [reply recommender/engine/rl_agent/action_selector.py:21]: I have to wait for this PR then

File: recommender/engine/rl_agent/action_selector.py [hunk 3]
```diff
@@ -28,16 +29,13 @@ def __init__(self, K: int, service_embedder: Optional[nn.Module] = None) -> None
...
+            service_embedder or load_last_module(SERVICES_AUTOENCODER).encoder
         )
         self.forbidden_indices = self._get_service_indices(
             retrieve_forbidden_service_ids()
         )
-        assert (
-            self.I - len(self.forbidden_indices) >= self.K
-        ), "there aren't enough allowed services to recommend"
+        if self.I - len(self.forbidden_indices) >= self.K:
+            raise TooSmallRecommendationSpace
```
Comment (JanKapala) [inline recommender/engine/rl_agent/action_selector.py:38]: I have changed this to InsufficientRecommendationSpace (as you suggested) in PR #82 - so it should be refactored during merging and later refactoring

**Review by JanKapala — APPROVED** [review]

**Pull request merged**